### PR TITLE
Copy/paste REPL clipboard format

### PR DESCRIPTION
**Pull request by cston**
Copy all selected text to clipboard for Text and RTF formats and add a REPL format that contains the entire selection but differentiates inputs, outputs, and prompts. Paste uses the REPL format if it exists and pastes inputs and outputs but not prompts.

### CODE REVIEW

**Comment by cston** [comment]
@tmat, @dinov, @zooba, @amcasey, @kevinh-ms, @ManishJayaswal, @genlu

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 1]
```diff
@@ -2357,19 +2363,52 @@ private static NormalizedSnapshotSpanCollection GetSelectionSpans(ITextView text
...
+                {
+                    GetTextBlocks(blocks, span);
+                }
+                return BufferBlock.Serialize(blocks.ToArray());
+            }
+
+            private void GetTextBlocks(List<BufferBlock> blocks, SnapshotSpan span)
             {
                 // Find the range of source spans that cover the span.
                 var sourceSpans = GetSourceSpans(span.Snapshot);
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: All spans will have the same snapshot, correct?  Would it be better to pass in the list of source spans?

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 2]
```diff
@@ -584,6 +584,12 @@ public bool Paste()
                 {
                     InsertCode(format);
                 }
+                else if (Clipboard.ContainsData(ReplFormat))
+                {
+                    var blocks = BufferBlock.Deserialize((string)Clipboard.GetData(ReplFormat));
+                    var text = GetTextFromTextBlocks(blocks);
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Shouldn't each input be pasted separately?  What if one is a REPL command (e.g. ```#help``` or ```#clear```)?

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 3]
```diff
@@ -2388,28 +2426,59 @@ private StringBuilder GetTextWithoutPrompts(StringBuilder builder, SnapshotSpan
...
-                        {
+                    }
+                    if (!added)
+                    {
+                        break;
+                    }
+                }
+            }
+
+            private const string InputBlockKind = "input";
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Can we use ```ReplSpanKind``` instead of strings?

Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Does there need to be a kind for standard input?

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 4]
```diff
@@ -2388,28 +2426,59 @@ private StringBuilder GetTextWithoutPrompts(StringBuilder builder, SnapshotSpan
...
+                    var mappedSpans = TextView.BufferGraph.MapDownToBuffer(span, SpanTrackingMode.EdgeExclusive, sourceSnapshot.TextBuffer);
+                    bool added = false;
+                    foreach (var mappedSpan in mappedSpans)
                     {
-                        var sourceSnapshot = sourceSpan.Snapshot;
-                        var mappedSpans = TextView.BufferGraph.MapDownToBuffer(span, SpanTrackingMode.EdgeExclusive, sourceSnapshot.TextBuffer);
-                        bool added = false;
-                        foreach (var mappedSpan in mappedSpans)
+                        var intersection = sourceSpan.Span.Intersection(mappedSpan);
+                        if (intersection.HasValue && !intersection.Value.IsEmpty)
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Are empty intersections dropped to solve a problem?

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 5]
```diff
@@ -2357,19 +2363,52 @@ private static NormalizedSnapshotSpanCollection GetSelectionSpans(ITextView text
                 return selection.SelectedSpans;
             }
 
+            /// <summary>
+            /// REPL clipboard format.
+            /// </summary>
+            private const string ReplFormat = "REPL";
```
Comment (DinoV) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Should this be InteractiveCode or something along those lines?  We don't use REPL anywhere else

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 1 above)
Comment (cston) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: All spans should be from the same snapshot, although we don't control how the selection was set. Will leave as is for now.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 5 above)
Comment (cston) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Changed to a GUID as @tmat suggested.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 4 above)
Comment (cston) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Empty intersections should only occur at the end of the session (input prompt but no input), and the resulting block will be ignored by paste.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 3 above)
Comment (cston) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Used ReplSpanKind and included StandardInput.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 2 above)
Comment (cston) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Makes sense. Changed.

**Comment by cston** [comment]
@dotnet-bot retest this please

**Pull request merged**